File: kattis/addingwords.cc
// https://open.kattis.com/problems/addingwords
#include <iostream>
#include <sstream>
#include <unordered_map>

using namespace std;

typedef unordered_map<string, int> msi;
typedef unordered_map<int, string> mis;


int main() {
    msi a;
    mis b;
    string l;
    while (getline(cin, l)) {
        stringstream in(l);
        string c;
        in >> c;
        if (c == "def") {
            string n;
            int v;
            in >> n >> v;
            if (a.find(n) != a.end()) b.erase(a[n]);
            a[n] = v;
            b[v] = n;
        } else if (c == "calc") {
            int r = 0;
            bool result = true;
            string n;
            in >> n;
            if (a.find(n) == a.end()) result = false;
            else r = a[n];
            cout << n << " ";
            while (true) {
                in >> n;
                cout << n << " ";
                if (n == "=") break;
                bool plus = n == "+";
                in >> n;
                cout << n << " ";
                if (a.find(n) == a.end()) result = false;
                else if (plus) r += a[n];
                else r -= a[n];
            }
            if (!result || b.find(r) == b.end()) cout << "unknown\n";
            else cout << b[r] << endl;
        } else if (c == "clear") {
            a.clear();
            b.clear();
        }
    }
}

File: kattis/addingwords.py
# https://open.kattis.com/problems/addingwords
a = {}
b = {}
while True:
    try:
        line = raw_input()
    except:
        break
    fs = line.split()
    if fs[0] == 'def':
        n = fs[1]
        v = int(fs[2])
        if n in a: del(b[a[n]])
        a[n] = v
        b[v] = n
    elif fs[0] == 'calc':
        r = 0
        result = True
        if fs[1] not in a: result = False
        else: r = a[fs[1]]
        for i in range(2, len(fs), 2):
            if fs[i] == '=': break
            plus = fs[i] == '+'
            if fs[i + 1] not in a: result = False
            elif plus: r += a[fs[i + 1]]
            else: r -= a[fs[i + 1]]
        print ' '.join(fs[1:]),
        if not result or r not in b: print 'unknown'
        else: print b[r]
    elif fs[0] == 'clear':
        a = {}
        b = {}
